docs/designs/keras_callback.md

```diff
@@ -0,0 +1,299 @@
+
+### Use a Container for Callbacks Defined in the Model
+We may define several callbacks for the model in a job. Tensorflow creates a container `CallbackList` to wrap the callbacks to conveniently call the methods in callbacks. For example:
+```python
+class CallbackList():
+    def on_batch_end(self, batch, logs=None):
+        for callback in self.callback_list:
+            callback.on_batch_end(bach, logs)
+```
+For detail, we can view the [source code](https://github.com/tensorflow/tensorflow/blob/cf7fcf164c9846502b21cebb7d3d5ccf6cb626e8/tensorflow/python/keras/callbacks.py#L189-L196
+) of `CallbackList` in Tensorflow.
+
+So, we can also use CallbackList to wrap the callbacks in ElasticDL.
+```python
+from tensorflow.python.keras.callbacks import CallbackList
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.on_batch_end(batch)
+```
+
+### Set Default Attributes for Callbacks.
+There are `set_model` and `set_params` in `tf.keras.callbacks.Callback`. `set_model` can set a model object to the attribute `model` of callback and `set_params` can set a dict object to the attribute `params` of callback.
+We also can use those methods of `CallbackList` to set `model` and `params` for the callbacks in `CallbackList`. 
+```python
+model = custom_model()
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.set_model(model)
+
+params = {
+    'batch_size': batch_size,
+    'epochs': epochs,
+    'saved_model_path': saved_model_path,
+    'checkpoint_path': checkpoint_path
+}
+callback_list.set_params(params)
+```
+
+Then, we can call `model` and `params` in the methods like:
+```python
+class CustomCallback(tf.keras.callbacks.Callback):
+    def on_train_batch_begin(self, batch, logs=None):
+        lr = self.model.optimizer.learning_rate,
+        saved_model_path = self.params.get("save_model_path")
+```
+
+
+## Implement Callbacks to Support the Cases in the Motivation
+We split the callbacks to two parts. One is the pre-configured callback which is automatically configured in ElasticDL and users don't need to define it in the model definition. And another is pre-made callback which users can configure in the model definition if needed.
+
+### Pre-configured Callback in ElasticDL
+
+* Case 5: Callback to export model using SavedModel after the training is completed.
+```python
+class SaveModelExporter(tf.keras.callbacks.Callback):
+    """Export model using SavedModel after training
+    Args:
+        task_data_service: TaskDataService to process data according the task
+        dataset_fn: function to process dataset
+        model_handler: to transform the trained model with ElasticDL embedding layer to Keras native model.
+    """
+    def __init__(self, task_data_service, dataset_fn, model_handler):
+        self._model_handler = model_handler
+        self._task_data_service
+        self._dataset_fn = dataset_fn
+
+    def on_train_end(self, logs=None):
```

Comment:
So these are ElasticDL codes, not user-defined callback codes?

Probably use different styles for ElasticDL codes and user-defined callback codes to make it more clear.

Reply:
Yes, `SavedModelExporter` is ElasticDL code. And we will provide some pre-made user-defined callback where users only need to define the logic like `LearningRateScheduler`. But the methods of all callbacks are the same, such as `on_train_end` and so on. What's the mean of "different styles"?

Reply:
"different styles" means markdown formats.